File: solutions/javascript/src/algorithm/others/Exercise7.js
import DynamicSetADT from "../../structure/dynamic/DynamicSetADT.js";
import DynamicSimpleDictionaryADT from "../../structure/dynamic/DynamicSimpleDictionaryADT.js";
import DynamicGraphADT from "../../structure/dynamic/DynamicGraphADT.js";

function calcularGradoEntradaSalida(grafo) {
  const resultado = new DynamicSimpleDictionaryADT(); // Diccionario simple para guardar resultados

  // Copiamos los vértices para poder iterar y modificar sin problemas
  const vertices = grafo.getVertxs();
  const tempSet = new DynamicSetADT();

  // Extraer todos los vértices a tempSet e inicializar resultado con [0, 0]
  while (!vertices.isEmpty()) {
    const v = vertices.choose();
    vertices.remove(v);
    tempSet.add(v);
    resultado.add(v, [0, 0]); // [gradoEntrada, gradoSalida]
  }

  // Restaurar vertices
  while (!tempSet.isEmpty()) {
    const v = tempSet.choose();
    tempSet.remove(v);
    vertices.add(v);
  }

  // Calcular grados
  // gradoSalida: contar aristas que salen de v
  // gradoEntrada: contar aristas que llegan a v
  for (const v of verticesToArray(vertices)) {
    let salida = 0;
    for (const edge of grafo.edges) {
      if (edge.from === v) {
        salida++;
        // actualizar grado de entrada del destino
        if (resultado.get(edge.to)) {
          const [entrada, salidaActual] = resultado.get(edge.to);
          resultado.remove(edge.to);
          resultado.add(edge.to, [entrada + 1, salidaActual]);
        } else {
          resultado.add(edge.to, [1, 0]);
        }
      }
    }
    // actualizar grado salida para v
    if (resultado.get(v)) {
      const [entrada, _] = resultado.get(v);
      resultado.remove(v);
      resultado.add(v, [entrada, salida]);
    } else {
      resultado.add(v, [0, salida]);
    }
  }

  return resultado;
}

// Función auxiliar para convertir DynamicSetADT a array sin modificar el conjunto
function verticesToArray(dynamicSet) {
  const result = [];
  const temp = new DynamicSetADT();

  while (!dynamicSet.isEmpty()) {
    const val = dynamicSet.choose();
    dynamicSet.remove(val);
    result.push(val);
    temp.add(val);
  }
  // Restaurar
  while (!temp.isEmpty()) {
    const val = temp.choose();
    temp.remove(val);
    dynamicSet.add(val);
  }

  return result;
}


const grafo = new DynamicGraphADT();
grafo.addVertx(1);
grafo.addVertx(2);
grafo.addVertx(3);

grafo.addEdge(1, 2, 5);
grafo.addEdge(1, 3, 10);
grafo.addEdge(2, 3, 7);

const grados = calcularGradoEntradaSalida(grafo);

const keys = grados.getKeys();
while (!keys.isEmpty()) {
  const k = keys.choose();
  keys.remove(k);
  const [entrada, salida] = grados.get(k);
  console.log(`Nodo ${k} -> Grado entrada: ${entrada}, Grado salida: ${salida}`);
}

File: solutions/python/algorithm/others/Exercise7.py
from structure.dynamic.DynamicGraphADT import DynamicGraphADT
from structure.dynamic.DynamicSetADT import DynamicSetADT
from structure.dynamic.DynamicSimpleDictionaryADT import DynamicSimpleDictionaryADT


def calcular_grado_entrada_salida(grafo: DynamicGraphADT):
    resultado = DynamicSimpleDictionaryADT()
    vertices = grafo.get_vertices()  # DynamicSetADT
    temp = DynamicSetADT()

    grados = {}

    # Extraer todos los vértices para inicializar grados
    while not vertices.isEmpty():
        nodo = vertices.choose()
        vertices.remove(nodo)
        temp.add(nodo)
        grados[nodo] = [0, 0]  # [grado entrada, grado salida]

    # Restaurar el conjunto original
    while not temp.isEmpty():
        nodo = temp.choose()
        temp.remove(nodo)
        vertices.add(nodo)

    # Recorremos cada vértice para contar salidas e incrementar entradas
    current_vertex = grafo.first_vertex
    while current_vertex is not None:
        origen = current_vertex.value
        # Contar salidas
        edge = current_vertex.edges
        salida_count = 0
        while edge is not None:
            salida_count += 1
            destino = edge.destination
            if destino not in grados:
                grados[destino] = [0, 0]
            grados[destino][0] += 1  # Incrementar entrada para destino
            edge = edge.next
        grados[origen][1] = salida_count  # Grado salida del nodo origen
        current_vertex = current_vertex.next

    # Cargar los resultados en un diccionario simple
    for nodo in grados:
        entrada, salida = grados[nodo]
        resultado.add(nodo, (entrada, salida))

    return resultado




grafo = DynamicGraphADT()
grafo.add_edge(1, 2, 10)
grafo.add_edge(2, 3, 5)
grafo.add_edge(1, 3, 2)

dic = calcular_grado_entrada_salida(grafo)
dic.mostrar()

# Salida esperada:
# Clave: 1 -> Valor: (0, 2)
# Clave: 2 -> Valor: (1, 1)
# Clave: 3 -> Valor: (2, 0)
